store/src/java/com/zimbra/cs/mailclient/MailConnection.java

```diff
@@ -80,24 +81,37 @@ public synchronized void connect() throws IOException {
             processGreeting();
             switch (config.getSecurity()) {
             case TLS:
-                startTls();
+                try {
+                    startTls();
+                }
+                catch (SSLHandshakeException e) {
+                    throw new IOException("STARTTLS failed. Please check the SSL certificate on the external server.");
+                }
                 break;
             case TLS_IF_AVAILABLE:
                 try {
                     startTls();
                 } catch (CommandFailedException e) {
                     getLogger().debug("STARTTLS failed", e);
                 }
+                catch (SSLHandshakeException e) {
+                    throw new IOException("STARTTLS failed. Please check the SSL certificate on the external server.");
+                }
                 break;
             }
-        } catch (IOException e) {
+        }
+        catch (SSLHandshakeException e) {
+            throw new IOException("Encrypted connection failed. Please check the SSL certificate on the external server.");
```

Comment:
My view on Gordon's question is that we should only need to catch `SSLHandshakeException` here.
Please log the exception at least at debug level to aid operator trouble shooting if needed.

Reply:
Currently the system is logging the entire stacktrace related to where this IOException originates anyhow.
I don't feel like that adding additional logging is useful at this point unless I can find a way to suppress the stacktrace itself.


Looks like this:

    857534-209:https:https://192.168.56.101/service/soap/TestDataSourceRequest:1516740769414:a9be13b9b025ab62
    E
    ra.common.service.ServiceException.FAILURE(ServiceException.java:288)
    ra.cs.datasource.Pop3Sync.connect(Pop3Sync.java:160)
    ra.cs.datasource.Pop3Sync.test(Pop3Sync.java:112)
    ra.cs.datasource.DataSourceManager.test(DataSourceManager.java:267)
    ra.cs.service.mail.TestDataSource.testDataSource(TestDataSource.java:152)
    ra.cs.service.mail.TestDataSource.handle(TestDataSource.java:52)
    ra.soap.SoapEngine.dispatchRequest(SoapEngine.java:623)
    ra.soap.SoapEngine.dispatch(SoapEngine.java:476)
    ra.soap.SoapEngine.dispatch(SoapEngine.java:274)
    ra.soap.SoapServlet.doWork(SoapServlet.java:303)
    ra.soap.SoapServlet.doPost(SoapServlet.java:213)
    rvlet.http.HttpServlet.service(HttpServlet.java:707)
    ra.cs.servlet.ZimbraServlet.service(ZimbraServlet.java:206)
    rvlet.http.HttpServlet.service(HttpServlet.java:790)
    pse.jetty.servlet.ServletHolder.handle(ServletHolder.java:821)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1685)
    ra.cs.servlet.CsrfFilter.doFilter(CsrfFilter.java:169)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.RequestStringFilter.doFilter(RequestStringFilter.java:54)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.SetHeaderFilter.doFilter(SetHeaderFilter.java:59)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ETagHeaderFilter.doFilter(ETagHeaderFilter.java:47)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ContextPathBasedThreadPoolBalancerFilter.doFilter(ContextPathBasedThreadPoolBalancerFilter.java:107)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ZimbraQoSFilter.doFilter(ZimbraQoSFilter.java:116)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    ra.cs.servlet.ZimbraInvalidLoginFilter.doFilter(ZimbraInvalidLoginFilter.java:117)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    pse.jetty.servlets.DoSFilter.doFilterChain(DoSFilter.java:473)
    pse.jetty.servlets.DoSFilter.doFilter(DoSFilter.java:318)
    pse.jetty.servlets.DoSFilter.doFilter(DoSFilter.java:288)
    pse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668)
    pse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581)
    pse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
    pse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
    pse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
    pse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1158)
    pse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511)
    pse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    pse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1090)
    pse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    pse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213)
    pse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:109)
    pse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    pse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:318)
    pse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:437)
    pse.jetty.server.handler.DebugHandler.handle(DebugHandler.java:84)
    pse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:119)
    pse.jetty.server.Server.handle(Server.java:517)
    pse.jetty.server.HttpChannel.handle(HttpChannel.java:306)
    pse.jetty.server.HttpConnection.onFillable(HttpConnection.java:242)
    pse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
    pse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    pse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:192)
    pse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:261)
    pse.jetty.io.FillInterest.fillable(FillInterest.java:95)
    pse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:75)
    pse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:213)
    pse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:147)
    pse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
    pse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
    g.Thread.run(Thread.java:748)
    IOException: STARTTLS failed. Please check the SSL certificate on the external server
    ra.cs.mailclient.MailConnection.connect(MailConnection.java:99)
    ra.cs.datasource.Pop3Sync.connect(Pop3Sync.java:152)

Reply:
If I'm reading that correctly, that is logging the ServiceException which has wrapped the new IOException that this pull request creates - thus the original SSLHandshakeException is still lost.  So, I would recommend logging both but perhaps only at debug level.  It can sometimes be useful to log just the message for an exception if the log level is info or above but be more detailed for debug or below.  It kind of looks like the ServiceException in this case doesn't have a useful message - is that possible?  If so, that seems sub-optimal too.

Reply:
Yes that is correct..  The actual exception is a CertificateException but that is actually buried inside the SSLException it seems.  It's like a russian doll of exceptions.